cmd/thanos/rule.go

```diff
@@ -472,49 +490,22 @@ func runRule(
 	}
 
 	// Handle reload and termination interrupts.
-	reload := make(chan struct{}, 1)
 	{
 		cancel := make(chan struct{})
-		reload <- struct{}{} // Initial reload.
-
 		g.Add(func() error {
+			// Initialize rules.
+			if err := reloadRules(logger, ruleFiles, ruleMgr, evalInterval, metrics); err != nil {
+				level.Error(logger).Log("msg", "initialize rules failed", "err", err)
+			}
 			for {
 				select {
-				case <-cancel:
-					return errors.New("canceled")
-				case <-reload:
 				case <-reloadSignal:
-				}
-
-				level.Debug(logger).Log("msg", "configured rule files", "files", strings.Join(ruleFiles, ","))
-				var files []string
-				for _, pat := range ruleFiles {
-					fs, err := filepath.Glob(pat)
-					if err != nil {
-						// The only error can be a bad pattern.
-						level.Error(logger).Log("msg", "retrieving rule files failed. Ignoring file.", "pattern", pat, "err", err)
-						continue
+					if err := reloadRules(logger, ruleFiles, ruleMgr, evalInterval, metrics); err != nil {
+						level.Error(logger).Log("msg", "reload rules by sighup failed", "err", err)
 					}
-
-					files = append(files, fs...)
-				}
-
-				level.Info(logger).Log("msg", "reload rule files", "numFiles", len(files))
-
-				if err := ruleMgr.Update(evalInterval, files); err != nil {
-					configSuccess.Set(0)
-					level.Error(logger).Log("msg", "reloading rules failed", "err", err)
-					continue
-				}
-
-				configSuccess.Set(1)
-				configSuccessTime.SetToCurrentTime()
-
-				rulesLoaded.Reset()
-				for _, group := range ruleMgr.RuleGroups() {
-					rulesLoaded.WithLabelValues(group.PartialResponseStrategy.String(), group.File(), group.Name()).Set(float64(len(group.Rules())))
+				case <-cancel:
+					return errors.New("canceled")
```

Comment:
I know it's not yours, but can we move to `context.WithCancel(..)` as we do everywhere else?